pkg/scheduler/controller/scheduler_control.go

```diff
@@ -0,0 +1,320 @@
+	c.reportSuccessfulScheduling(shoot, seed.Name)
+	return nil
+}
+
+// determineSeed returns an appropriate Seed cluster (or nil).
+func determineSeed(shoot *gardenv1beta1.Shoot, seedLister gardenlisters.SeedLister, shootLister gardenlisters.ShootLister, strategy config.CandidateDeterminationStrategy) (*gardenv1beta1.Seed, error) {
+	seedList, err := seedLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	shootList, err := shootLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	return determineBestSeedCandidate(shoot, shootList, seedList, strategy)
+}
+
+func determineBestSeedCandidate(shoot *gardenv1beta1.Shoot, shootList []*gardenv1beta1.Shoot, seedList []*gardenv1beta1.Seed, strategy config.CandidateDeterminationStrategy) (*gardenv1beta1.Seed, error) {
+	// Map seeds to number of managed shoots.
+	var (
+		seedUsage  = generateSeedUsageMap(shootList)
+		candidates []*gardenv1beta1.Seed
+	)
+
+	switch strategy {
+	case config.SameRegion:
+		candidates = determineCandidatesWithSameRegionStrategy(seedList, shoot, candidates)
+	case config.MinimalDistance:
+		candidates = determineCandidatesWithMinimalDistanceStrategy(seedList, shoot, candidates)
+	default:
+		return nil, fmt.Errorf("unknown seed determination strategy configured in seed admission plugin. Strategy: '%s' does not exist. Valid strategies are: %v", strategy, config.Strategies)
+	}
+
+	if candidates == nil {
+		message := fmt.Sprintf("no matching seed found for Configuration (Cloud Profile '%s', Region '%s', SeedDeterminationStrategy '%s')", shoot.Spec.Cloud.Profile, shoot.Spec.Cloud.Region, strategy)
+		return nil, errors.New(message)
+	}
+
+	old := candidates
+	candidates = nil
+
+	for _, seed := range old {
+		if hasDisjointedNetworks := validateDisjointedNetworks(seed, shoot); hasDisjointedNetworks {
+			candidates = append(candidates, seed)
+		}
+	}
+
+	if candidates == nil {
+		return nil, errors.New("found %d possible seed cluster(s), however none have a disjoint network")
+	}
+
+	var (
+		bestCandidate *gardenv1beta1.Seed
+		min           *int
+	)
+
+	// Find the best candidate (i.e. the one managing the smallest number of shoots right now).
+	for _, seed := range candidates {
+		if numberOfManagedShoots := seedUsage[seed.Name]; min == nil || numberOfManagedShoots < *min {
+			bestCandidate = seed
+			min = &numberOfManagedShoots
+		}
+	}
+
+	return bestCandidate, nil
+}
+
+func determineCandidatesWithSameRegionStrategy(seedList []*gardenv1beta1.Seed, shoot *gardenv1beta1.Shoot, candidates []*gardenv1beta1.Seed) []*gardenv1beta1.Seed {
+	// Determine all candidate seed clusters matching the shoot's cloud and region.
+	for _, seed := range seedList {
+		if seed.DeletionTimestamp == nil && seed.Spec.Cloud.Profile == shoot.Spec.Cloud.Profile && seed.Spec.Cloud.Region == shoot.Spec.Cloud.Region && seed.Spec.Visible != nil && *seed.Spec.Visible && verifySeedAvailability(seed) {
```

Comment:
The Seeds are being retrieved with the Lister.

So it see two options: 
1) use the client always to fetch from API Server. Needs changes in the API Server Strategy.
2) build a Lister Expansion that does the filtering for us like [here](https://github.com/kubernetes/client-go/blob/7b18d6600f6b0022e31c46b46875beffd85cc71a/listers/core/v1/node_expansion.go#L34). However this would only push the filtering logic to the Lister instead

Reply:
Offline discussion: we keep the current implementation.